### DISCUSSION TITLE
16-bit color strips

### DISCUSSION THREAD

**Discussion by r3vEngIT** [awawa-dev/HyperHDR#526] [Ideas]
For many years I used @dmadison set-up with Prismatik fork. This obviously had its limitations and after switching to Windows 11 and HDR I decided to use your system. I had many issues at first but have worked through [most] of them. I appreciate the immense amount of time you have devoted to this and share openly with us; therefore, my suggestions are only to be of help.

Current set-up:
Rpi 2b v1.1, SD card version of HyperHDR
HD107s 60/m strip, 284 total LED
Flatbuffers protocol @30fps, switch between Rpi and Nvidia card

I was running 144/m strip for a total of 510 LED with no lag but my 1000w PSU was not enough and on full white scenes the Rpi would crash and reboot (I am powering the Rpi off the strip via GPIO pins back through the same cable that I am sending SPI over and while not recommended, with 60/m strip i have had no issues).

I first used a grabber via zone2 HDMI out of my AVR but the lag was high (lacking usb3 on the Pi) and the AVR did not like configuration and would consistently kick out HDR in Windows. I did not want to use a splitter because even a good one will process the signal. My goal is passthrough of 2160p HDR10 Dolby Atmos 7.1. 

I then used HyperHDR as the master in Windows to an Arduino Nano Every using SPI and adalight. Then I tried HyperHDR forwarder to Rpi to Nano Every. Both worked but had frequent connection and communication issues with Windows10 and 11 made that all worse. I also had too much cabling between my HTPC and TV. I will say, the Nano Every is very capable if the user wants to still use Arduino.

Then I found @sabaatworld HyperionScreenCap. This was it. I could use the same Ethernet cable already used to communicate with HyperHDR on Rpi and using shielded cat8 cable I had strong data transfer over longer distances. I have no latency issues, the colors that are sent to the Rpi are exactly what the Windows screen is displaying so no having to mess with calibrations, and the connection is automatic when turning on my HTPC. 

So my question is, why are you using a ESP32 connected to the Rpi instead of communicating with the LED strip directly from the Rpi GPIO? Can the Rpi not do parallel pwm output as well? 

Some issues that the HD107s strip has eliminated for me:
1. No need for a 5v level shifter. pixel chip designed for 3.3v.
2. less voltage drop, and less power injection points due to thicker copper pcb
3. no parallel output needed, i have had 600 pixels with no latency at 60fps with data only at one end.
4. colors are much better than ws2812b strip and produces very bright white. 
5. no need to use smoothing

I have never used an RGB/W strip. is the white color reproduction that much better than the HD107 to justify all the other devices, wiring, programming, etc that must be done for the R/W strip?

My next goal is to use the 16-bit color HD108 strip but I am not skilled enough like you are to make it work. So I have an offer:
I had to have a big order manufactured in order to get these strips. I have been re-selling them but if I gave you HD108 60/m strips to keep for free... would you be able to make them work with your HyperHDR for Rpi?

**Comment by r3vEngIT** [comment]
Here is my current setup:

![PXL_20221224_071944432](https://user-images.githubusercontent.com/35564353/222930783-a79bae9a-517f-4c4a-a1bc-b42800b6be23.jpg)

![PXL_20230304_220310465](https://user-images.githubusercontent.com/35564353/222930682-dcb8019d-cfe6-4cc8-88dd-93eefb68f2d8.jpg)



https://user-images.githubusercontent.com/35564353/222931360-49fa3914-725c-4faa-9ebc-7dde98fb811d.mp4

**Comment by r3vEngIT** [comment]
additional suggestions:

instead of wood, using aluminum angle is stronger and acts like a heat sink for the LEDs. if you use the alum angle that is 1"x2" it creates an angle off the back of the TV less than 45 degrees which is IMHO a better light projection on the wall. 

something that I really liked about Prismatik were the color adjustments. adjusting the LED strip based on Kelvin scale color temp is much more intuitive than adjusting color 0-255 scale individually. 

the ability to change modes from color backlight to dark scenes with a single click was great too. It took me a while to dial this in with HyperHDR and because i had to mess with so many parameters to do it, some color reproduction has suffered.

**Comment by ghost** [comment]
Good afternoon. Could you take a picture of the back of your TV. I'm interested to see what the aluminum profile looks like.

<details>
  <summary>Found this profile:</summary>
  
  ![frame](https://user-images.githubusercontent.com/85223482/222953403-f800741c-d20e-4d73-9534-bb8a492b4542.jpg)

</details>

There is matte plastic and transparent.

But I still don’t understand how to attach it on TV, it still has its own weight along with the tape.
I don't know if it will hold securely with double sided tape (the profile will heat up from the tape, distributing heat, how it will affect the glue of this tape).

**Reply by r3vEngIT** [reply]
i tried a LED channel like that before because i thought the diffusing would make the light projection look better but the matte plastic used by the manufactures is not very good. the best diffusing of LED strips ive seen is done with silicone (sold as "neon pixel rope"). 

the aluminum i was referring to was common stock sold at a hardware store. to attach to the tv i have tried lots of different methods using a combination of duct tape, hot glue and tip ties. i am always making changes to my set up, however, and needed something more "modular." 

my first design used magnets. neodymium ones with holes in them. i found a spot on the back of the tv i could safely drill a hole to rivet the magnet to the tv and then did the same to the aluminum frame i built for the strips. the frame i built used aluminum angle stock. i held it together by cutting the corners at 45 degrees and drilling tiny holes at those ends to use a small piece of alum to rivet at the corner like a bracket. 
this worked well but the frame was not very sturdy and the magnets a pita to line up and pull apart. one side of the frame also had to be assembled on the tv too otherwise it would not fit around the tv (my tv is mounted to the wall with a custom mount i made that houses all the wires and power supplies.)

my current design is pretty clever :) 
i used the aluminum frame from a window screen. this is very light but sturdy. the corners are held together with slips that make for easy disassembly and allow you to cut the frame to any length you need and slip back together. the frame is slightly bigger than the width of the strip and hollow for great heat dissipation and if you wanted to you could run the wires used for power injection inside the frame for a nice clean setup. 

to mount it i got lucky and my tv has an area that protrudes out which allowed me to rest a cross brace on. all i have to do now is lower the frame on the back of the tv. the brace holds the top and sides together and then i can slip the bottom frame and strip into the sides. easy breezy. for the data in i soldered a usb port. i was experimenting with so many different controllers in the beginning that all the soldering would destroy the pads. usb cable has the perfect wires (+, d, d, -) for a LED strip and i had lots of cables i could cut up and put on the controllers to allow easy connecting and disconnecting. 

hope this helps.

![PXL_20230304_220241847](https://user-images.githubusercontent.com/35564353/223042387-11c21c3d-bd07-4cbd-b8eb-db2dc9ef5ad3.jpg)
![PXL_20230306_061559038](https://user-images.githubusercontent.com/35564353/223042551-906941a7-1d49-40e8-97b2-940ae89d8862.jpg)

**Reply by ghost** [reply]
> hope this helps

Thanks for the answer. After looking at your TV, I now understand how I can attach the tape.
Unlike you, my back side of the TV is not straight, but slightly curved at the edges.
<details>
  <summary> </summary>
  
  ![static lg](https://user-images.githubusercontent.com/85223482/223050924-5aa81f3f-99cc-4e0e-a4e0-6c16218446c0.png)

</details>
I will make the same general horizontal attachment point in the middle of the TV.
The TV has factory holes in which the screws are hidden. They are quite deep and I can easily place a plastic dowel in them.
<details>
  <summary> </summary>
  
  ![images](https://user-images.githubusercontent.com/85223482/223047785-29904aa5-4c88-420c-a92d-cf766dcf5058.jpg)

</details>

And on it I can fix the central aluminum profile. And don't have to make any holes in the TV.
And on this central axis I can attach an already frame with a LED tape.
Thus, the frame will lag behind the TV body a little and its rounded shapes will not be a hindrance.

Thanks for your very helpful advice!!!

**Reply by r3vEngIT** [reply]
you are welcome.

it is hard to see in the picture but my TV has curved edges too. i also have the frame lagging behind the TV because of this. 

another reason i used the window screen frame was because it was free! haha
![PXL_20230306_061644789](https://user-images.githubusercontent.com/35564353/223061498-7224aa4f-33e9-428b-9cf0-b82d3cc8d9d1.jpg)

**Reply by AstaRom** [reply]
> it is hard to see in the picture but my TV has curved edges too

From this angle, you can see it well :)

**Comment by awawa-dev** [comment]
Hi

> I was running 144/m strip for a total of 510 LED with no lag but my 1000w PSU was not enough and on full white scenes the Rpi would crash and reboot (I am powering the Rpi off the strip via GPIO pins back through the same cable that I am sending SPI over and while not recommended, with 60/m strip i have had no issues).

That's why I always recommend a dedicated USB power supply for the Raspberry Pi. I'm not surprised that a voltage spike with such a large configuration can cause voltage instability during sudden scene/LED brightness changes. You also bypass the protection of the Rpi by powering it through the pinout instead of using the USB port. What is even more dangerous is that the power supply is using an unstable LED power supply, whose characteristics change under a sudden load because it also powers the LEDs.

> So my question is, why are you using a ESP32 connected to the Rpi instead of communicating with the LED strip directly from the Rpi GPIO? Can the Rpi not do parallel pwm output as well?

99% of such users setups don't use a mandatory level shifter. Neopixel protocol requires strict timings, emulation (it is not native SPI protocol like for APA102, it needs a special data buffer to be prepared first to simulate it), Rpi solution usually starts to malfunction around ~250LEDs (occasional blinking) sometimes it starts around 200LEDs especially if RGBW LED strip is used. With larger setups it simply doesn't work (usually last segment doesn't work at all or is constantly blinking). 
However, I have nothing against controlling the SPI LEDs from the Raspberry Pi, as long as the level shifter is used.

> instead of wood, using aluminum angle is stronger and acts like a heat sink for the LEDs. if you use the alum angle that is 1"x2" it creates an angle off the back of the TV less than 45 degrees which is IMHO a better light projection on the wall.

Agree, I use it mainly for cost-effectiveness, and also wood is easier to work with. I still have no problems with heating (96 diodes / meter, diodes and its surroundings on the strip have the ambient temperature to the touch). My guess is that the temperature problem may arise with a very large setup, but it's not the LEDs themselves under normal conditions/use, only with the internal copper LED tracks if only one power injection is used, causing high current flow using only one path. However, it is still a problem of bad design, which such cooling would only mask. A 45 angle was also used because my frame is very close to the wall, about 3 cm. But for a greater distance, the angle should be smaller (the angle of incidence/reflection, but also the angle/intensity distribution of light scattering on the material/wall towards the observer)

> No need for a 5v level shifter. pixel chip designed for 3.3v.

I could not find a HD107s specification that could confirm that. Always the high logic level voltage specification is missing. If it works for you then OK, but for 5V LEDs strip I stick to the 3.3V to 5V level shifter recommendation. For example for Neopixel LED strip like WS2812b high logic level is defined usually as 0.7 * Vdd. I know there are WS2812b versions that can handle logic voltage as low as ~2.8V, only problem is to recognize such versions. Since we are not so sure, we also have to assume the use of a level shifter. 

> colors are much better than ws2812b strip and produces very bright white.

The high density ws2812b can also produce very bright white. But so what if the white produced from the RGB components has a much worse purity than the white produced by RGBW ... in addition, it changes depending on the white scale and is impossible to calibrate.

> My next goal is to use the 16-bit color HD108 strip but I am not skilled enough like you are to make it work. So I have an offer:
I had to have a big order manufactured in order to get these strips. I have been re-selling them but if I gave you HD108 60/m strips to keep for free... would you be able to make them work with your HyperHDR for Rpi?

This is HD108 protocol? https://www.rose-lighting.com/wp-content/uploads/sites/53/2019/11/HD108-led.pdf It does not require much work to be able to control such a LED strip. I can implement it even blindly in HyperHDR without sample of this LED, but then you would have to test it yourself. But we will not use it to its full potential for now (LED is 16bits, all inputs/grabbers work at 8bits).

> I have never used an RGB/W strip. is the white color reproduction that much better than the HD107 to justify all the other devices, wiring, programming, etc that must be done for the R/W strip?

I used RGB ws2801 for years, ws2812b temporarily and still have ws2813 as one of my test setup. Yes, RGBW is much better and I can't imagine going back to RGB LEDs, even if they offered 16 bits of resolution. sk6812 is not perfect due to the slow neopixel data protocol, but only with the HD108 RGBW version I would consider an upgrade ... they have been announcing them for years, but nothing has come of it, and as you noticed, there is a problem even with buying such a 16-bit RGB version, so probably they focus on solutions for companies, not consumers.

> something that I really liked about Prismatik were the color adjustments. adjusting the LED strip based on Kelvin scale color temp is much more intuitive than adjusting color 0-255 scale individually.

Solution for color temperature is only available for RGBW in HyperHDR for some LED drivers. This is done outside, not in HyperHDR, and is used to correct the characteristics of a given type of LED: they already have a factory-defined temperature, e.g. in the case of "neutral" sk6812 RGBW, we must cut off a large share of red and green so that the white color is actually white and not yellowish . I do not anticipate similar solutions for RGB for the reasons I wrote about: we are not able to provide software workaround for similar quality as with RGBW diodes.

**Reply by r3vEngIT** [reply]
i was mistaken, i did stop injecting power to the Rpi GPIO. now i have the wires soldered directly to the usb pads on the underside of the board with a capacitor. i am also using a computer PSU for stable power.

Question: for months i had the Rpi controlling 510 LEDs with no level shifter, yet, did not experienced any signs (like flickering). 
but could i be losing performance without knowing it? if the first LED will bump up the signal to 5V then wouldnt the second LED be the one to flicker? why would the end of the strip flicker?

i had a real bad issue with flickering one time. it just randomly started one day. i never completely figured it out but it was with a pro micro that has 5V logic. my conclusion is that it had something to do with controller speed. this is another reason i switched to hyperhdr. i like how you have it set up to easily change the baud rate and how you have the protocol for each LED chip type already configured. using Arduino ide every time i wanted to make a change was a pita lol.
it took me a while to get good at configuring hyperhdr but there is much more i like about it vs prismatik.

suggestion: the notification banner that pops up when saving a change stays up too long and is too big. i find myself having to force close it bc it will be in the way of the next thing i want to look at. especially if multiple changes are made causing the banners to stack on the screen haha. can this be improved? 

Question: does the smoothing feature have any benefit if using 4-wire LED strip? i do not have flickering issues with still images but would FPS improve with it on? if yes, is it worth the performance impact it has on the Rpi? 

i understand what you mean with White color scale. i cannot dim the LEDs very much or it will cause the white color to not match the screen. even at full, i can notice the strip's white not being able to perfectly match the screen's white. this is my main reason for wanting to use the 16-bit strips. having more steps in the color scale will allow for smoother and more accurate color changes when lowering the brightness vs 8-bit. 

Question: Yes, that is the correct HD108 from roselighting. you mention not being used to its full potential because of other limiting factors such as the grabber. but what if not using a grabber? is flatbuffer protocol 8-bit? 
my GPU, AVR and TV all sync with 12-bit color & Full dynamic range. would this cause the 8-bit strip to not reproduce colors correctly? 

that would be wonderful if you could implement the HD108 and i will definitely test it out :)

and if you change your mind and would like a HD108 strip i will still give you one for all your great work.

**Reply by awawa-dev** [reply]
> Question: for months i had the Rpi controlling 510 LEDs with no level shifter, yet, did not experienced any signs (like flickering).
but could i be losing performance without knowing it? if the first LED will bump up the signal to 5V then wouldnt the second LED be the one to flicker? why would the end of the strip flicker?

Theoretically, it is possible if the high signal becomes temporarily unrecognizable at the LEDs due to voltage spikes in their power supply then even first LED amplifier wont help (usually a voltage lower than 5v on the LED strip is conducive to the fact that the 3.3v signal is recognized as 'high'). When the LED strip see only low signal then usually it wont refresh/update. 

> suggestion: the notification banner that pops up when saving a change stays up too long and is too big. i find myself having to force close it bc it will be in the way of the next thing i want to look at. especially if multiple changes are made causing the banners to stack on the screen haha. can this be improved?

Yes, it annoys me too. Will do something about it.

> Question: does the smoothing feature have any benefit if using 4-wire LED strip? i do not have flickering issues with still images but would FPS improve with it on? if yes, is it worth the performance impact it has on the Rpi?

I don't see the benefit of smoothing above 100Hz, and due to the specificity of the QTimer in the system (which is used by smoothing), higher refresh rates tend to have rather unstable intervals.

> Yes, that is the correct HD108 from roselighting. you mention not being used to its full potential because of other limiting factors such as the grabber. but what if not using a grabber? is flatbuffer protocol 8-bit?
my GPU, AVR and TV all sync with 12-bit color & Full dynamic range. would this cause the 8-bit strip to not reproduce colors correctly?

Currently all sources including flatbuffer are using only 8bit scale for color components. 

I found updated HD108 documentation here: [http://www.rose-lighting.com/wp-content/uploads/sites/53/2021/03/HD108-5050-LED-Specificaion-V1.1.0-Rose-Lighting.pdf](https://www.rose-lighting.com/wp-content/uploads/sites/53/2021/03/HD108-5050-LED-Specificaion-V1.1.0-Rose-Lighting.pdf)

This LED strip is not compatible with 3.3V logic level. It may work with 3.3V under favorable conditions, but it doesn't have to and it will certainly be out of specification.
![obraz](https://user-images.githubusercontent.com/69086569/223283863-a339bd5b-3636-41ed-a363-5cada485038a.png)

---

@r3vEngIT 
I created a PR that brings support for HD108: https://github.com/awawa-dev/HyperHDR/pull/527
You can download the archive with the installer here: https://github.com/awawa-dev/HyperHDR/actions/runs/4379283313
If you can test it and give feedback if it works for you. `Adaptive Current Threshold` currently work only to dim the LED (1000 => 100%).

**Reply by r3vEngIT** [reply]
I am going to have to get the HD108s setup on my TV to test. I am also going to get a level shifter ordered and wired in. 

It will take me a few days to complete but I will test and provide feedback ASAP.

Thank you!

**Reply by r3vEngIT** [reply]
you may already be aware of this but i found some other works with the HD108 that may be useful:

https://github.com/tinic/hd108-calibration
https://www.reddit.com/r/FastLED/comments/mod8mn/fastled_branch_with_16bit_support_hd108/

**Comment by r3vEngIT** [comment]
i found where i read about these strips being 3.3v tolerant. it is on the Newstar site. the rose site (including the data sheet) is a mess of copy and pasted information. 
https://www.hd108-led.com/how-to-driver-hd108-leds/
![image](https://user-images.githubusercontent.com/35564353/224356343-07509922-de09-4765-86d6-6e42a2842bc2.png)


this data sheet may be better to use:
http://newstar-ledstrip.com/product/20220705143016452.pdf

i was also reading that CMOS chips, which are what these strips use, are traditionally 3.3v logic.
